BaseClasses.py

```diff
@@ -516,6 +516,49 @@ def get_spheres(self) -> Iterator[Set[Location]]:
                 state.collect(location.item, True, location)
             locations -= sphere
 
+    def get_sendable_spheres(self) -> Iterator[Set[Location]]:
+        """
+        yields a set of multiserver sendable locations (location.item.code: int) for each logical sphere
+
+        If there are unreachable locations, the last sphere of reachable locations is followed by an empty set,
+        and then a set of all of the unreachable locations.
+        """
+        state = CollectionState(self)
+        locations = set()
+        events = set()
+        for location in self.get_filled_locations():
+            if type(location.item.code) == int:
+                locations.add(location)
+            else:
+                events.add(location)
+
+        while locations:
+            sphere: Set[Location] = set()
+
+            # cull events out
+            done_events = {None}
+            while done_events:
+                done_events = set()
+                for event in events:
+                    if event.can_reach(state):
+                        state.collect(event.item, True, event)
+                        done_events.add(event)
+                events -= done_events
+
+            for location in locations:
+                if location.can_reach(state):
+                    sphere.add(location)
+
+            yield sphere
+            if not sphere:
+                if locations:
+                    yield locations  # unreachable locations
+                break
+
+            for location in sphere:
+                state.collect(location.item, True, location)
```

Comment:
I know there has been some confusion about the usage of the word `event` in this code base.
The 2nd parameter to this `collect` function is named `event`.
That makes this look wrong, but I think it's probably right.

Someone who understands the confusion better than I do should double-check it.
(It could be out of scope for this PR, but maybe this parameter name should have been renamed to `advancement` with a change that happened a few months ago. https://github.com/ArchipelagoMW/Archipelago/pull/2871)

Reply:
i agree with renaming it but `advancement` is wrong here. What this param does is when true, it forces the item to be collected into state, and prevents a sweep. the first part here seems like it might cause some issues down the line but i'm not entirely sure where all this is done that it'd actually be relevant

Reply:
it's a copy from right above.